engine/src/sellst.cpp

```diff
@@ -98,30 +119,59 @@ bool MCSellist::getids(MCListRef& r_list)
 	return MCListCopy(*t_list, r_list);
 }
 
+void MCSellist::Clean()
+{
+    if (objects == nil)
+        return;
+    
+    // Remove any dead objects from the selected list
+    MCSelnode* t_cursor = objects;
+    do
+    {
+        if (!t_cursor->m_ref)
+        {
+            MCSelnode* t_next = t_cursor->next();
+            t_cursor->remove(objects);
+            delete t_cursor;
+            t_cursor = t_next;
+        }
+        else
+        {
+            t_cursor = t_cursor->next();
+        }
+    }
+    while (t_cursor != objects);
+}
+
 void MCSellist::clear(Boolean message)
 {
 	if (locked)
 		return;
-	MCObject *optr = NULL;
-	while (objects != NULL)
+    
+	MCObjectHandle optr = nil;
+	while (objects != nil)
 	{
 		MCSelnode *nodeptr = objects->remove(objects);
-		optr = nodeptr->ref;
+		optr = nodeptr->m_ref;
 		delete nodeptr;
 	}
+    
 	MCundos->freestate();
-	if (message && optr != NULL)
+    
+	if (message && optr)
 		optr->message(MCM_selected_object_changed);
 }
 
 void MCSellist::top(MCObject *objptr)
 {
-	if (objects != NULL)
+	Clean();
```

Comment:
@livecodefraser Could you please refactor this patch so that the list is only cleaned when the list is being mutated anyway?